### ISSUE TITLE
feat: Add differential privacy (DP-SGD) for training-time privacy guarantees

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#27]
## Summary

Implement differential privacy for training pipelines to protect training data from inference attacks.

## Current State

The Sovereign AI Stack has **artifact-level** encryption:
- `.apr` format (aprender) - Model file encryption (AES-256-GCM)
- `.ald` format (alimentar) - Dataset file encryption (AES-256-GCM)

**Missing:** Training-time privacy guarantees (DP-SGD).

## Threat Model

Without DP, trained models can leak training data through:
- **Membership inference attacks** - Determine if a sample was in training set
- **Model inversion attacks** - Reconstruct training data from model
- **Training data extraction** - Extract memorized sequences (especially LLMs)

### DP-SGD (Differentially Private Stochastic Gradient Descent)

```rust
use entrenar::{DpSgd, PrivacyBudget, GradientClipper};

// Configure privacy budget
let budget = PrivacyBudget::new()
    .epsilon(1.0)           // Privacy loss bound
    .delta(1e-5)            // Failure probability
    .target_epochs(10);     // Compute noise for N epochs

// Training with DP-SGD
let optimizer = DpSgd::new(AdamW::default())
    .with_budget(budget)
    .with_max_grad_norm(1.0)    // Gradient clipping
    .with_noise_multiplier(1.1); // Calibrated noise

pipeline
    .with_optimizer(optimizer)
    .train(&dataset)?;

// Query remaining budget
println!("Spent: ε={:.2}", optimizer.spent_epsilon());
```

### Features

- [ ] **Gradient clipping** - Per-sample gradient norm bounding
- [ ] **Gaussian noise injection** - Calibrated to privacy budget
- [ ] **Privacy accounting** - RDP (Rényi DP) composition
- [ ] **Budget tracking** - Query spent ε at any time
- [ ] **Auto-calibration** - Compute noise multiplier from (ε, δ, epochs)

### Privacy Accounting

```rust
use entrenar::privacy::{RdpAccountant, compute_noise_multiplier};

// Compute required noise for target privacy
let noise = compute_noise_multiplier(
    epsilon: 1.0,
    delta: 1e-5,
    sample_rate: 0.01,  // batch_size / dataset_size
    epochs: 10,
)?;

// Or track with RDP accountant
let mut accountant = RdpAccountant::new();
for epoch in 0..10 {
    train_epoch(&mut model, &optimizer);
    accountant.step(noise_multiplier, sample_rate);
}
let (eps, delta) = accountant.get_privacy_spent(target_delta: 1e-5);
```

## References

- [DP-SGD Paper](https://arxiv.org/abs/1607.00133) - Abadi et al., 2016
- [Opacus](https://github.com/pytorch/opacus) - PyTorch DP library (reference impl)
- [TensorFlow Privacy](https://github.com/tensorflow/privacy)
- [RDP Accounting](https://arxiv.org/abs/1702.07476) - Mironov, 2017

## Part of Sovereign Privacy Stack

| Layer | Protection | Status |
|-------|------------|--------|
| alimentar (.ald) | Dataset encryption | ✅ Complete |
| aprender (.apr) | Model encryption | ✅ Complete |
| entrenar (DP-SGD) | Training privacy | This issue |

**Comment by noahgift** [comment]
Implemented in src/optim/dp/ with full DP-SGD support including privacy accountant (accountant.rs), privacy budget tracking (budget.rs), gradient clipping/noise (gradient.rs), and comprehensive tests.
